File: src/python/reverse_words_in_a_string.py
"""
Question
https://www.educative.io/courses/grokking-coding-interview-patterns-python/reverse-words-in-a-string

Given a sentence, reverse the order of its words without affecting the order of
letters within the given word.

Note: The input string may contain leading or trailing spaces or multiple
spaces between words. The returned string, however, should only have a single
space separating each word. Do not include any extra spaces.

#two-pointers
"""


import re


def reverse_words(sentence) -> str:
    sentence = re.sub(" +", " ", sentence.strip())
    sentence = list(sentence)
    strLen = len(sentence)

    reverse(sentence, 0, strLen - 1)
    left = 0
    right = 0

    while left < strLen:
        while right < strLen and sentence[right] != " ":
            right += 1
        reverse(sentence, left, right - 1)
        left = right + 1
        right += 1

    return "".join(sentence)


def reverse(sentence, left, right):
    while left < right:
        sentence[left], sentence[right] = sentence[right], sentence[left]
        left += 1
        right -= 1

File: src/javascript/reverse_words_in_a_string.js
/*
Question
https://www.educative.io/courses/grokking-coding-interview-patterns-python/reverse-words-in-a-string

Given a sentence, reverse the order of its words without affecting the order of
letters within the given word.

Note: The input string may contain leading or trailing spaces or multiple
spaces between words. The returned string, however, should only have a single
space separating each word. Do not include any extra spaces.

#two-pointers
*/


function reverse_words(sentence) {
  sentence = sentence.trim().replace(/\s+/g, " ");
  sentence = sentence.split("");
  const strLen = sentence.length;
  reverse(sentence, 0, strLen - 1);

  let left = 0;
  let right = 0;

  while (left < strLen) {
    while (right < strLen && sentence[right] !== " ") {
      right++;
    }
    reverse(sentence, left, right - 1);
    left = right + 1;
    right++;
  }

  return sentence.join("");
}

function reverse(arr, left, right) {
  while (left < right) {
    const temp = arr[left];
    arr[left] = arr[right];
    arr[right] = temp;

    left++;
    right--;
  }
}
